dbms/src/Flash/Mpp/MPPTaskManager.h

```diff
@@ -54,6 +54,35 @@ struct MPPQueryTaskSet
     }
 };
 
+/// A simple thread unsafe FIFO cache used to fix the "lost cancel" issues
+class AbortedMPPGatherCache
+{
+private:
+    std::deque<MPPGatherId> gather_ids;
+    std::unordered_set<MPPGatherId, MPPGatherIdHash> gather_ids_set;
+    size_t capacity;
+
+public:
+    AbortedMPPGatherCache(size_t capacity_)
+        : capacity(capacity_)
+    {}
+    bool exists(const MPPGatherId & id)
+    {
+        return gather_ids_set.find(id) != gather_ids_set.end();
+    }
+    void add(const MPPGatherId & id)
+    {
+        if (gather_ids_set.size() >= capacity && gather_ids_set.find(id) == gather_ids_set.end())
```

Comment:
It looks like L81 also needs to check `gather_ids_set.find(id) == gather_ids_set.end()` first
```suggestion
        if (gather_ids_set.find(id) == gather_ids_set.end())
            return;
        if (gather_ids_set.size() >= capacity)
```

Reply:
Or it is only written in this way, the duplicate id will also enter the cache, but the existing entry will not be eliminated?

Reply:
Seems duplicated gather id will cause duplicate ids in gather_ids, a little strange